dao/src/main/java/com/epam/esm/repository/GiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,222 @@
+		CriteriaQuery<GiftCertificate> criteriaQuery = criteriaBuilder.createQuery(GiftCertificate.class);
+		Root<GiftCertificate> root = criteriaQuery.from(GiftCertificate.class);
+		root.fetch("tags", JoinType.LEFT);
+		criteriaQuery.select(root).where(criteriaBuilder.equal(root.get("id"), id));
+		GiftCertificate gift = session.createQuery(criteriaQuery).getResultStream().findFirst().orElse(null);
+		session.close();
+		return gift;
+	}
+
+	@Override
+	public List<GiftCertificate> findByAnyParams(Long size, String substr, Integer skip, Integer limit, String sort,
+												 String order) {
+		Session session = sessionFactory.openSession();
+		CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+		CriteriaQuery<GiftCertificate> criteriaQuery2 = criteriaBuilder.createQuery(GiftCertificate.class);
+		Root<GiftCertificate> giftCertificateRoot = criteriaQuery2.from(GiftCertificate.class);
+		Predicate finalPredicate = getFinalPredicate(size, substr, session, criteriaBuilder, giftCertificateRoot);
+		if (finalPredicate != null) {
+			criteriaQuery2.select(giftCertificateRoot).where(finalPredicate).groupBy(giftCertificateRoot.get("id"));
+		}
+		orderByFieldName(sort, order, criteriaBuilder, criteriaQuery2, giftCertificateRoot);
+		List<GiftCertificate> finalList = getPagingList(skip, limit, session, criteriaQuery2);
+		session.close();
+		return finalList;
+	}
+
+	@Override
+	public Double findPriceById(Integer id) {
+		Session session = sessionFactory.openSession();
+		CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+		CriteriaQuery<Double> criteriaQuery = criteriaBuilder.createQuery(Double.class);
+		Root<GiftCertificate> root = criteriaQuery.from(GiftCertificate.class);
+		criteriaQuery.select(root.get("price")).where(criteriaBuilder.equal(root.get("id"), id));
+		Double price = session.createQuery(criteriaQuery).getResultStream().findFirst().orElse(null);
+		session.close();
+		return price;
+	}
+
+	@Override
+	public String findNameById(Integer id) {
+		Session session = sessionFactory.openSession();
+		CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+		CriteriaQuery<String> criteriaQuery = criteriaBuilder.createQuery(String.class);
+		Root<GiftCertificate> root = criteriaQuery.from(GiftCertificate.class);
+		criteriaQuery.select(root.get("name")).where(criteriaBuilder.equal(root.get("id"), id));
+		String name = session.createQuery(criteriaQuery).getResultStream().findFirst().orElse(null);
+		session.close();
+		return name;
+	}
+
+	@Override
+	public int update(Map<String, Object> updates, Integer id, Instant now) {
+		Session session = sessionFactory.openSession();
+		CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+		CriteriaUpdate<GiftCertificate> criteriaUpdate = criteriaBuilder.createCriteriaUpdate(GiftCertificate.class);
+		Root<GiftCertificate> root = criteriaUpdate.from(GiftCertificate.class);
+		updates.forEach((key, value) -> {
+			criteriaUpdate.set(key, value);
```

Comment:
please take a look for Idea suggestions. 

you can find ALL idea suggestions via menu Code -> Inspect code. Not all advice are really useful, but some of them can help to improve the code or find a bug